File: apps/mobile/lib/sync/live.ts
// Live-sync channel: a WebSocket to GET /v1/sync/live whose ONLY job is to
// tell us WHEN to pull. On connect the SERVER subscribes the socket to every
// vault the account is an active member of (no client-side vault list), then
// sends {"t":"poke","vault_id":...} whenever a vault gains committed events.
// The channel carries NO ledger data — a poke just triggers the existing
// authenticated pull, which is cursor-idempotent (a self-poke after our own
// push pulls nothing). The scheduler's polling stays untouched as the
// fallback whenever this socket is down.
//
// Wire contract (mirrored server-side):
//   - Auth: session JWT. Primary "Authorization: Bearer <jwt>" header (React
//     Native's WebSocket supports a headers option); "?token=<jwt>" query
//     param as fallback. Header wins server-side.
//   - Server → client: {"t":"poke","vault_id":"<uuid>"} and {"t":"ping"}
//     (keepalive every 30s).
//   - Client → server: {"t":"pong"} in reply to ping — nothing else. The
//     server drops the conn after 2 unanswered pings.
//
// Lifecycle: start()/stop() from the scheduler (foreground + signed-in only).
// Reconnect backoff 1s → 2 → 5 → 10 → 30s cap, jittered, reset on a
// successful open. Never throws into callers — every failure path is a
// silent retry (plus __DEV__ logs).

import { getBackendUrl } from "../api";

const BACKOFF_STEPS_MS = [1_000, 2_000, 5_000, 10_000, 30_000];
// ±25% jitter so a fleet dropped by one server restart doesn't redial in
// lockstep.
const JITTER_FRACTION = 0.25;

// React Native's WebSocket takes a third options arg with custom headers
// (how we attach Authorization). The ambient TS declaration only types the
// two-arg DOM form, so we go through this narrowed constructor type.
type RNWebSocketCtor = new (
  url: string,
  protocols?: string | string[] | null,
  options?: { headers?: Record<string, string> } | null,
) => WebSocket;

export type LiveChannelOpts = {
  // Reads the current session JWT (null while signed out — treated as a
  // retryable condition, so a later sign-in gets picked up by the backoff
  // loop without a restart).
  getJwt: () => Promise<string | null>;
  // Called for every poke. Must not block: do debounce/coalesce work inside.
  // Exceptions are swallowed here so a listener bug can't kill the socket.
  onPoke: (vaultId: string) => void;
};

export type LiveChannel = {
  start: () => void;
  stop: () => void;
};

export function connectLiveChannel(opts: LiveChannelOpts): LiveChannel {
  // start() called and stop() not yet — the retry loop only runs while true.
  let desired = false;
  let ws: WebSocket | null = null;
  let retryTimer: ReturnType<typeof setTimeout> | null = null;
  let attempt = 0;
  // Guards the async dial: stop()+start() while a dial's awaits are pending
  // bumps this, and the stale dial bails instead of racing the fresh one.
  let generation = 0;
  let dialing = false;

  const clearRetryTimer = (): void => {
    if (retryTimer) {
      clearTimeout(retryTimer);
      retryTimer = null;
    }
  };

  const scheduleRetry = (): void => {
    if (!desired || retryTimer) return;
    const base = BACKOFF_STEPS_MS[Math.min(attempt, BACKOFF_STEPS_MS.length - 1)];
    attempt++;
    // Jitter in [1 - f, 1 + f) around the slot.
    const delay = Math.round(base * (1 - JITTER_FRACTION + Math.random() * 2 * JITTER_FRACTION));
    retryTimer = setTimeout(() => {
      retryTimer = null;
      void dial();
    }, delay);
  };

  // Detach handlers BEFORE close so the dying socket's onclose can't schedule
  // a retry after stop() (or fight a fresh socket after a restart).
  const teardownSocket = (): void => {
    const sock = ws;
    if (!sock) return;
    ws = null;
    try {
      sock.onopen = null;
      sock.onmessage = null;
      sock.onerror = null;
      sock.onclose = null;
      sock.close();
    } catch {
      // Already closed / never connected — nothing to do.
    }
  };

  const handleMessage = (data: unknown): void => {
    try {
      if (typeof data !== "string") return;
      const msg = JSON.parse(data) as { t?: string; vault_id?: string };
      if (msg?.t === "ping") {
        // Keepalive: the server drops us after 2 unanswered pings.
        ws?.send(JSON.stringify({ t: "pong" }));
      } else if (msg?.t === "poke" && typeof msg.vault_id === "string" && msg.vault_id) {
        opts.onPoke(msg.vault_id);
      }
      // Unknown t → ignore (forward-compatible with new server frames).
    } catch (err) {
      if (__DEV__) console.warn("[sync.live] message ignored", err);
    }
  };

  const dial = async (): Promise<void> => {
    if (!desired || ws || dialing) return;
    dialing = true;
    const gen = ++generation;
    try {
      let jwt: string | null = null;
      try {
        jwt = await opts.getJwt();
      } catch {
        jwt = null;
      }
      if (!desired || gen !== generation) return;
      if (!jwt) {
        // Signed out: retry on the backoff schedule (one cheap SecureStore
        // read per slot) so the channel comes up after the next sign-in.
        scheduleRetry();
        return;
      }

      const baseUrl = await getBackendUrl();
      if (!desired || gen !== generation) return;
      // http → ws, https → wss; the poke endpoint is a WebSocket upgrade.
      const wsBase = baseUrl.replace(/^http/, "ws").replace(/\/+$/, "");
      // ?token= is the fallback for stacks that strip the Authorization
      // header off upgrade requests; the server prefers the header.
      const url = `${wsBase}/v1/sync/live?token=${encodeURIComponent(jwt)}`;

      const sock = new (WebSocket as unknown as RNWebSocketCtor)(url, null, {
        headers: { Authorization: `Bearer ${jwt}` },
      });
      ws = sock;
      sock.onopen = () => {
        // Connected — future drops start the backoff ladder from the bottom.
        attempt = 0;
        if (__DEV__) console.log("[sync.live] connected");
      };
      sock.onmessage = (e) => handleMessage(e.data);
      sock.onerror = () => {
        // RN always follows onerror with onclose; reconnect is handled there.
      };
      sock.onclose = () => {
        ws = null;
        if (__DEV__) console.log("[sync.live] closed, will retry");
        scheduleRetry();
      };
    } catch (err) {
      // Constructor / URL resolution failure — same retry path as a drop.
      if (__DEV__) console.warn("[sync.live] dial failed", err);
      if (ws) teardownSocket();
      scheduleRetry();
    } finally {
      dialing = false;
    }
  };

  const start = (): void => {
    if (desired) return; // already connected or retrying
    desired = true;
    attempt = 0;
    void dial();
  };

  const stop = (): void => {
    desired = false;
    generation++; // invalidate any dial that's mid-await
    clearRetryTimer();
    attempt = 0;
    teardownSocket();
  };

  return { start, stop };
}

File: apps/backend/internal/sync/live.go
// Live-sync poke channel — GET /v1/sync/live (WebSocket).
//
// The socket carries NO ledger data. It exists only to collapse the polling
// latency between "co-editor pushed" and "my device pulled": the server sends
// {"t":"poke","vault_id":"…"} when a vault gains committed events (or its
// membership changes), and the client reacts by running the SAME
// authenticated pull it would have run on its next poll tick. Every
// authorization decision therefore still happens on the pull path — a poke
// received by a just-revoked member leads to a 403, never to data.
//
// Because polling continues to exist as the backstop, delivery here is
// deliberately best-effort: pokes are dropped rather than ever blocking
// PushEvents, and a dead socket is simply reaped (the client reconnects).
package sync

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	gosync "sync"
	"time"

	"github.com/coder/websocket"

	"github.com/matee/kaata-backend/internal/auth"
	"github.com/matee/kaata-backend/internal/httpx"
)

const (
	// livePokeBuffer is each subscriber's poke-channel capacity. When a
	// burst overflows it, Notify DROPS the poke instead of blocking —
	// harmless, because (a) any one queued poke already means "pull this
	// vault", extra ones are redundant, and (b) the client's regular poll
	// still exists as the backstop for a missed poke.
	livePokeBuffer = 8

	// maxLiveSocketsPerAccount is a leak guard, not a feature limit. A
	// shopkeeper has a handful of devices; more than 8 concurrent sockets
	// for one account means a client-side reconnect loop that isn't closing
	// its old connections. Rejecting with 429 pre-upgrade keeps the server's
	// goroutine count bounded by 8 × accounts.
	maxLiveSocketsPerAccount = 8

	// liveMaxMissedPings: the conn is dropped when this many consecutive
	// pings go unanswered by a {"t":"pong"}.
	liveMaxMissedPings = 2

	// liveWriteTimeout bounds every frame write. A peer that stopped
	// reading (dead radio, suspended app) would otherwise pin the writer
	// goroutine on a full TCP buffer forever — after hijack the http.Server
	// Write/Idle timeouts no longer apply, so the socket must police itself.
	liveWriteTimeout = 10 * time.Second
)

// livePingInterval implements the keepalive contract: {"t":"ping"} every 30s.
// JSON-level pings (not WS protocol pings) because React Native's WebSocket
// does not expose protocol-level ping/pong to application code. A var, not a
// const, ONLY so live_test.go can shrink it — production code never writes it.
var livePingInterval = 30 * time.Second

// liveMsg is the single wire shape for both directions.
//
//	server → client: {"t":"poke","vault_id":"<uuid>"} | {"t":"ping"}
//	client → server: {"t":"pong"}
type liveMsg struct {
	T       string `json:"t"`
	VaultID string `json:"vault_id,omitempty"`
}

// liveSub is one connected socket's subscription. The vault set is resolved
// ONCE at connect time (see Handler.Live) — a vault joined mid-connection
// starts poking after the client's next reconnect, and a vault revoked
// mid-connection keeps receiving (harmless) pokes until disconnect. Both
// staleness windows are acceptable because pokes carry no data and the pull
// path re-checks membership on every request.
type liveSub struct {
	accountID string
	ch        chan string // buffered vault_id pokes
}

// liveBroker is the in-memory per-vault fanout. Single-instance deploy —
// same assumption as every other in-process cache in this Service (membership
// LRU, pull-page LRU): one backend replica, so no Redis/pubsub indirection.
type liveBroker struct {
	mu gosync.Mutex
	// subs is vault_id → the sockets subscribed to it. One liveSub appears
	// under every vault in its connect-time membership list.
	subs map[string]map[*liveSub]struct{}
	// perAccount counts open sockets per account for the leak guard.
	perAccount map[string]int
}

func newLiveBroker() *liveBroker {
	return &liveBroker{
		subs:       make(map[string]map[*liveSub]struct{}),
		perAccount: make(map[string]int),
	}
}

// subscribe registers a socket for every vault in vaultIDs. Returns an error
// when the account already holds maxLiveSocketsPerAccount sockets — callers
// must reject BEFORE upgrading so the client sees a plain HTTP 429.
func (b *liveBroker) subscribe(accountID string, vaultIDs []string) (*liveSub, error) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if b.perAccount[accountID] >= maxLiveSocketsPerAccount {
		return nil, fmt.Errorf("account %s already holds %d live sockets", accountID, maxLiveSocketsPerAccount)
	}
	sub := &liveSub{
		accountID: accountID,
		ch:        make(chan string, livePokeBuffer),
	}
	for _, v := range vaultIDs {
		set, ok := b.subs[v]
		if !ok {
			set = make(map[*liveSub]struct{})
			b.subs[v] = set
		}
		set[sub] = struct{}{}
	}
	b.perAccount[accountID]++
	return sub, nil
}

// unsubscribe removes the socket from every vault set and releases its
// per-account slot. Idempotence is not needed — the single caller runs it
// exactly once via defer.
func (b *liveBroker) unsubscribe(sub *liveSub) {
	b.mu.Lock()
	defer b.mu.Unlock()
	for v, set := range b.subs {
		if _, ok := set[sub]; ok {
			delete(set, sub)
			if len(set) == 0 {
				delete(b.subs, v)
			}
		}
	}
	if n := b.perAccount[sub.accountID]; n <= 1 {
		delete(b.perAccount, sub.accountID)
	} else {
		b.perAccount[sub.accountID] = n - 1
	}
}

// notify fans a poke out to every subscriber of vaultID, INCLUDING the
// pusher's own socket — the pusher's follow-up pull is a cursor-idempotent
// no-op, and excluding it would require threading the pushing socket's
// identity through PushEvents for zero benefit. Non-blocking by
// construction: a full channel drops the poke (see livePokeBuffer).
func (b *liveBroker) notify(vaultID string) {
	b.mu.Lock()
	defer b.mu.Unlock()
	for sub := range b.subs[vaultID] {
		select {
		case sub.ch <- vaultID:
		default:
			// Subscriber's buffer is full: it already has pending pokes it
			// hasn't drained, so it will pull anyway. Dropping keeps this
			// call O(subscribers) with zero blocking on the push hot path.
		}
	}
}

// NotifyLive fans a poke out for vaultID. Exposed on the Service so the
// push commit path and the membership-invalidation entry point (both in
// service.go) can poke without knowing broker internals.
func (s *Service) NotifyLive(vaultID string) {
	if s.live != nil {
		s.live.notify(vaultID)
	}
}

// liveVaultIDs resolves the connect-time subscription set: every vault the
// account is an ACTIVE member of (accepted, not revoked). Archived vaults
// are deliberately NOT filtered out — the owner keeps sync access during the
// 30-day grace and should keep getting pokes; non-owners' pulls 403 with
// vault_archived, which is exactly the signal their client needs anyway.
func (s *Service) liveVaultIDs(ctx context.Context, accountID string) ([]string, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT vault_id::text FROM vault_members
		 WHERE account_id = $1::uuid
		   AND accepted_at IS NOT NULL
		   AND revoked_at IS NULL
	`, accountID)
	if err != nil {
		return nil, fmt.Errorf("live: vault list: %w", err)
	}
	defer rows.Close()
	var out []string
	for rows.Next() {
		var v string
		if err := rows.Scan(&v); err != nil {
			return nil, fmt.Errorf("live: vault list scan: %w", err)
		}
		out = append(out, v)
	}
	return out, rows.Err()
}

// LiveTokenFallbackMiddleware promotes ?token=<jwt> into the Authorization
// header when (and only when) no Authorization header is present — header
// wins. This lets the standard session middleware authenticate the WebSocket
// upgrade unchanged (signature, expiry, AND the 60s revocation cache) for
// clients whose WebSocket API cannot set headers. Register it BEFORE
// authenticator.Middleware() on the /v1/sync/live group only: promoting
// query tokens on ordinary REST routes would invite tokens into access logs
// and referrers for no reason.
func LiveTokenFallbackMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") == "" {
			if tok := r.URL.Query().Get("token"); tok != "" {
				r.Header.Set("Authorization", "Bearer "+tok)
			}
		}
		next.ServeHTTP(w, r)
	})
}

// Live — GET /v1/sync/live (PROTECTED; WebSocket upgrade).
func (h *Handler) Live(w http.ResponseWriter, r *http.Request) {
	claims, ok := auth.ClaimsFromContext(r.Context())
	if !ok {
		httpx.Error(w, http.StatusUnauthorized, "authentication required")
		return
	}

	vaultIDs, err := h.svc.liveVaultIDs(r.Context(), claims.AccountID)
	if err != nil {
		log.Printf("sync.live: vault list failed for account=%s: %v", claims.AccountID, err)
		httpx.Error(w, http.StatusInternalServerError, "live subscribe failed")
		return
	}

	// Reserve the subscription BEFORE upgrading so the per-account limit is
	// a clean HTTP 429 the client can back off on, not a post-upgrade close.
	sub, err := h.svc.live.subscribe(claims.AccountID, vaultIDs)
	if err != nil {
		httpx.Error(w, http.StatusTooManyRequests, "too many live connections for this account")
		return
	}
	defer h.svc.live.unsubscribe(sub)

	// OriginPatterns "*": this endpoint is consumed by the mobile app's
	// WebSocket client, which sends no browser Origin. The Origin check
	// exists to stop cross-site WebSocket hijacking of COOKIE-authenticated
	// endpoints; auth here is an explicit bearer JWT that a hostile web page
	// cannot obtain, so origin allowlisting adds nothing.
	conn, err := websocket.Accept(w, r, &websocket.AcceptOptions{
		OriginPatterns: []string{"*"},
	})
	if err != nil {
		// Accept already wrote the HTTP error response.
		return
	}
	// CloseNow is the unconditional cleanup; the graceful close paths below
	// run first when they apply.
	defer conn.CloseNow()

	// Own cancellation domain: after hijack the request context is no longer
	// cancelled by the server on connection loss, so the reader goroutine's
	// exit (via cancel) is what unwinds the writer loop.
	ctx, cancel := context.WithCancel(r.Context())
	defer cancel()

	// Reader: the client only ever sends {"t":"pong"}. Anything unparseable
	// is ignored rather than fatal — a lenient reader means a client-side
	// bug degrades to the ping-timeout path instead of a tight reconnect
	// loop. A read error (peer closed, network died) cancels the writer.
	pongCh := make(chan struct{}, 1)
	go func() {
		defer cancel()
		for {
			_, data, err := conn.Read(ctx)
			if err != nil {
				return
			}
			var m liveMsg
			if json.Unmarshal(data, &m) == nil && m.T == "pong" {
				select {
				case pongCh <- struct{}{}:
				default:
				}
			}
		}
	}()

	writeJSON := func(msg liveMsg) error {
		// Per-frame deadline; see liveWriteTimeout.
		wctx, wcancel := context.WithTimeout(ctx, liveWriteTimeout)
		defer wcancel()
		b, err := json.Marshal(msg)
		if err != nil {
			return err
		}
		return conn.Write(wctx, websocket.MessageText, b)
	}

	// Writer loop: pokes as they arrive, pings on the keepalive cadence.
	// awaitingPongs counts pings sent since the last pong; hitting
	// liveMaxMissedPings at the next tick means the peer answered neither of
	// the last two pings within 30s each — treat the conn as dead.
	pingTick := time.NewTicker(livePingInterval)
	defer pingTick.Stop()
	awaitingPongs := 0
	for {
		select {
		case <-ctx.Done():
			return
		case vaultID := <-sub.ch:
			if err := writeJSON(liveMsg{T: "poke", VaultID: vaultID}); err != nil {
				return
			}
		case <-pongCh:
			awaitingPongs = 0
		case <-pingTick.C:
			if awaitingPongs >= liveMaxMissedPings {
				// Graceful close so a half-alive peer that can still read
				// learns why; errors ignored — CloseNow follows regardless.
				_ = conn.Close(websocket.StatusPolicyViolation, "ping timeout")
				return
			}
			if err := writeJSON(liveMsg{T: "ping"}); err != nil {
				return
			}
			awaitingPongs++
		}
	}
}
